File: 82 Remove Duplicates from Sorted List II/deleteDuplicatesII.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        
        unordered_set<int> record;
        
        ListNode* dummyHead = new ListNode(0);
        dummyHead->next = head;
        ListNode* cur = dummyHead;
        ListNode* pre;
        
        while( cur->next != NULL ){
            
            // Record the node before all duplicate nodes
            if( record.find(cur->next->val) == record.end() )
                pre = cur;
            
            int flag = 0;
            while( cur->next != NULL && record.find(cur->next->val) != record.end() ){
                
                // Delete the next node
                ListNode* delNode = cur->next;
                cur->next = delNode->next;
                delete delNode;
                flag = 1;
            }
            
            // Delete the first node of the group of nodes with duplicate values, which is indicated by cur
            if(flag == 1){
                ListNode* delNode = cur;
                pre->next = cur->next;
                cur = pre;
                delete delNode;
            }
            
            if( cur->next == NULL )
                break;
            
            record.insert( cur->next->val );
            cur = cur->next;
        }
        
        ListNode* retNode = dummyHead->next;
        delete dummyHead;
        return retNode;
    }
};

File: 82 Remove Duplicates from Sorted List II/deleteDuplicatesII.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def deleteDuplicates(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        record = set()
        
        dummyHead = ListNode(0)
        dummyHead.next = head
        cur = dummyHead
        
        while cur.next != None:
            
            if cur.next.val not in record:
                pre = cur
            
            flag = 0
            while cur.next != None and cur.next.val in record:
                
                delNode = cur.next
                cur.next = delNode.next
                delNode.next = None
                flag = 1
                
            if flag == 1:
                
                delNode = cur
                pre.next = delNode.next
                cur = pre
                delNode.next = None
                
            if cur.next == None:
                break
                
            record.add(cur.next.val)
            cur = cur.next
        
        retNode = dummyHead.next
        dummyHead.next = None
        
        return retNode